src/ffi3/object.rs

```diff
@@ -466,6 +520,69 @@ mod typeobject {
         pub tp_next: *mut PyTypeObject,
     }
 
+    #[cfg(PyPy)]
```

Comment:
On train, I came up with a simpler way.
```rust
    macro_rules! _type_object_init {
        ({$($head:tt)*} $tp_as_async:ident, $($tail:tt)*) => {
            as_expr! {
                PyTypeObject {
                    $($head)*
                    tp_name: ptr::null(),
                    tp_basicsize: 0,
                    tp_itemsize: 0,
                    tp_dealloc: None,
                    tp_print: None,
                    tp_getattr: None,
                    tp_setattr: None,
                    $tp_as_async: ptr::null_mut(),
                    tp_repr: None,
                    tp_as_number: ptr::null_mut(),
                    tp_as_sequence: ptr::null_mut(),
                    tp_as_mapping: ptr::null_mut(),
                    tp_hash: None,
                    tp_call: None,
                    tp_str: None,
                    tp_getattro: None,
                    tp_setattro: None,
                    tp_as_buffer: ptr::null_mut(),
                    tp_flags: ffi3::object::Py_TPFLAGS_DEFAULT,
                    tp_doc: ptr::null(),
                    tp_traverse: None,
                    tp_clear: None,
                    tp_richcompare: None,
                    tp_weaklistoffset: 0,
                    tp_iter: None,
                    tp_iternext: None,
                    tp_methods: ptr::null_mut(),
                    tp_members: ptr::null_mut(),
                    tp_getset: ptr::null_mut(),
                    tp_base: ptr::null_mut(),
                    tp_dict: ptr::null_mut(),
                    tp_descr_get: None,
                    tp_descr_set: None,
                    tp_dictoffset: 0,
                    tp_init: None,
                    tp_alloc: None,
                    tp_new: None,
                    tp_free: None,
                    tp_is_gc: None,
                    tp_bases: ptr::null_mut(),
                    tp_mro: ptr::null_mut(),
                    tp_cache: ptr::null_mut(),
                    tp_subclasses: ptr::null_mut(),
                    tp_weaklist: ptr::null_mut(),
                    tp_del: None,
                    tp_version_tag: 0,
                    $($tail)*
                }
            }
        }
    }

    #[cfg(PyPy)]
    macro_rules! py_type_object_init {
        ($tp_as_async:ident, $($tail:tt)*) => {
            _type_object_init!({
                    ob_refcnt: 1,
                    ob_pypy_link: 0,
                    ob_type: ptr::null_mut(),
                    ob_size: 0,
                }
                $tp_as_async,
                tp_pypy_flags: 0,
                $($tail)*
            )
        }
    }

    #[cfg(not(PyPy))]
    macro_rules! py_type_object_init {
        ($tp_as_async:ident, $($tail:tt)*) => {
            _type_object_init!({
                ob_base: ffi3::object::PyVarObject {
                    ob_base: ffi3::object::PyObject_HEAD_INIT,
                    ob_size: 0
                },}
                $tp_as_async,
                $($tail)*
            )
        }
    }
```

Reply:
Cool, I think this highlights the differences in the head struct a bit better :)
Updated.